Shamima/fall2020/recruitment-challenge/submission

Shamima-Ali opened the pull request:
Implemented the Navigation class. 

**- def update_current_position(self)**
Used the PollSensor method from GPS to update the x, y and z values. The getPos method is then used to return the x, y and z values to the Navigation class.

**- def navigate(self)** 
Created 5 boolean variables that will store the boolean value returned from the Steering methods (move_up, move_down etc.).The boolean variables have been assigned a default value of True so if any of the Steering methods were unsuccessful, the variables will become False. If that happens, the TBM will stop moving and display an error message indicating which direction (x, y or z) it couldn't move to. The current_position will also be updated and compared to the desired_position.
------For moving in the x direction: As long as the desired_position is not equal to 0 or current_position, the TBM will move forwards.
------For moving in the y or z direction: The required y or z positions are found by subtracting the current_position and desired_position. If the required y or z directions are positive, the TBM will move to the right and upwards. It will move in the opposite directions if the required y or z positions are negative.
If all of the actuations are successful, the current_position will be updated and compared to the desired_position.

Commit:
Removed 'all' from README.md
git status

Shamima-Ali committed:
Implemented set_desired_position

Shamima-Ali committed:
Implemented update_current_position

Shamima-Ali committed:
Added code for navigate

Shamima-Ali committed:
Added comments

Burke-Daniel commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
     def update_current_position(self):
         """ Updates the current position of the TBM """
+        GPS.pollSensor
+        self.current_position = GPS.getPos
```
These are method calls, so the syntax will be:
```python
GPS.pollSensor()
self.current_position = GPS.getPos()
```

Burke-Daniel commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+        equal to the desired position in the x direction. """
+        if desired_position[0] != 0:
+            while(current_position[0] != desired_position[0]):
+                act_request_forward = Steering.move_forward()
```
so you are calling the methods in Steering as if they were static methods. In order to call these, you need to call them from the steering object that is owned by the Navigation class, so this would instead look like:
```python
act_request_forward = self.steering.move_forward()
```

colton-smith replied:
Steering is the class type, I think you mean to refer to the Steering instance that is a class field - "steering"

Burke-Daniel commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+                if act_request_forward == False:
+                    print("Unable to move to desired x position")
+                    self.update_current_position
+                    return current_position == desired_position
```
I like the way you have this, where if actuation fails it will return whether or not we have reached the desired position. Nifty way of doing it!

Burke-Daniel commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+
+        """"The TBM will move forward as long as the current position is not
+        equal to the desired position in the x direction. """
+        if desired_position[0] != 0:
```
should be `self.desired_position[0]`, as the desired_position variable is a member of the object

colton-smith commented:
So I don't forget, I'm gonna mention immediately that it would be a good idea to have the default "Actuation Successful" boolean values as False instead of True, for safety reasons.

colton-smith commented on README.md:
```diff
@@ -1,2 +1,3 @@
 # recruiting-challenges
-Repository containing all paradigm student group recruiting challenges.
+Repository containing paradigm student group recruiting challenges.
```
You're right, it does sound better without `all`.

Shamima-Ali replied:
I didn't really mean to make that correction😄   I was just trying out to 'commit' feature on my computer.

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+
+        self.desired_position = desired_position
+
         pass
```
All instances of pass should be removed, they were only placed there as placeholders.

Shamima-Ali replied:
Done

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
         Note: assume the user will pass in the desired_position parameter when using
         the interface 
         """
+
```
PEP8 - There should not be a blank line following any function/method docstring.

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+
         pass
 
     
```
PEP8 - Python methods should only be separated by a single blank line.

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+        act_request_down    = True
+
+
+        """"The TBM will move forward as long as the current position is not
```
There are 4 quotations here instead of 3

colton-smith replied:
Also, # is preferred for block comments as opposed to """

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
-        """
+        """ Assumption: If any actuation is unsuccessful, the TBM will stop moving"""
+
+        """ These boolean values will indicate whether or not the actuations were suucessful."""
```
The comment below navigate is a function docstring and should not have been removed, it acts as documentation for the function. It could have been modified, but the one you replaced it with doesn't describe the method.

It is good to include any assumptions in the docstring though (like preconditions).

Docstring notes/examples: https://www.python.org/dev/peps/pep-0257/

colton-smith replied:
Successful*

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+        act_request_right   = True
+        act_request_left    = True
+        act_request_up      = True
+        act_request_down    = True
```
In python it is uncommon to align things like this, instead a space is usually placed before and after the assignment. This is almost assembly-like. 

Ex:
``` number = 10 ```
``` position = 11 ```
``` velocity = 12 ```

colton-smith replied:
Good variable names

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
-        Returns: True if actuation requests were successful, False if not
-        Note: It may be good to notify the user if something unexpected happens!
-        """
+        """ Assumption: If any actuation is unsuccessful, the TBM will stop moving"""
```
I think this method is supposed to allow the calling code to stop the TBM in the event that an actuation is unsuccessful

colton-smith commented on fall-2020/navigation.py:
```diff
@@ -70,23 +70,68 @@ def set_desired_position(self, desired_position):
+        if required_y_distance > 0:
+            act_request_right = Steering.move_right(required_y_distance)
+        elif required_y_distance < 0:
+            act_request_left = Steering.move_left(required_y_distance)
```
Here, the required distance is negative, indicating you have to move left. However, you are calling move left and passing it this negative value. Depending on the implementation of the move_left method, this could steer the machine right or do nothing. Maybe you can use the abs method to get the magnitude of required_distance. 

This comment also applies to the z steering block.

colton-smith reviewed (changes requested):
Well done Shamima!

Git Practice:
Branch name - great, follows the requested format
PR Description - great, very descriptive. You could also mention the changes to the top level README.md just so that all your bases are covered and all files changed are noted.
PR Title - great, follows the requested format
Commit Messages - concise and descriptive

Code:
The code is very thorough and shows a lot of thought. There are a few minor things to address, and they are mostly style related.

Importantly:
- Steering is a type, you want to make sure you are using an instance of the class and not the type
- You are passing steer left and steer down negative values
- Pass is a placeholder and should be removed when the method is implemented

Thanks for the submission, you can effect some changes and re-request review if you so desire!

Burke-Daniel reviewed (changes requested):
I've just included a couple of nit-picky comments here, Colton got all the major things I noticed as well. Keep up the good work!

Shamima-Ali committed:
Update navigation.py

Thanks for taking the time to review the code. I made the changes as requested.
